scipy/special/_logsumexp.py

```diff
@@ -104,19 +104,43 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
     elif not np.isfinite(a_max):
         a_max = 0
 
+    tmp0 = (a == a_max) # compute largest separately
+
+
     if b is not None:
         b = np.asarray(b)
-        tmp = b * np.exp(a - a_max)
+        tmp = b * np.exp(a - a_max)  * (~tmp0)
+        tmp0 = b*(tmp0 + .0)
     else:
-        tmp = np.exp(a - a_max)
+        tmp = np.exp(a - a_max)  * (~tmp0)
+        tmp0 = tmp0 + .0
 
     # suppress warnings about log of zero
     with np.errstate(divide='ignore'):
         s = np.sum(tmp, axis=axis, keepdims=keepdims)
+        s0 = np.sum(tmp0, axis=axis, keepdims=keepdims)
+        sf = s + s0
         if return_sign:
-            sgn = np.sign(s)
+            sgn = np.sign(sf)
             s *= sgn  # /= makes more sense but we need zero -> zero
-        out = np.log(s)
+            s0*=sgn
+            sf*=sgn
+
+        if(np.isscalar(s0)):
```

Comment:
Best avoid using `np.isscalar`.  Not sure if it can be a scalar here at all, this comes from `np.sum(..., keepdims=True)`, so it's a 0D array at most.